### PR TITLE
docs: Add sphinxext.rediraffe redirection plugin

### LINKED ISSUE #11068: Broken documentation links
I've observed a number of broken links in Envoy documentation that previously worked (now showing a 404). Examples:

* https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/websocket (referenced from a google search of "envoy proxy web socket")
* https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log (referenced from other documentation on the internet)

It appears we're knowingly breaking these links without any sort of redirect to the new location. What's the policy for the Envoy site in this regard? Breaking existing links has an unfortunate chain-reaction effect on any product that relies on Envoy, and all Envoy users.

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: docs: Add sphinxext.rediraffe redirection plugin
Additional Description:

Adds a Sphinx extension that allows redirecting pages that have been moved, to ensure that old published links continue to work when docs change

Also has a tool for diffing against commit/branch to figure out which redirects should be added. This could potentially be added as part of CI, but i have not done so in this PR, as just fixing the links and adding the redirects seemed a greater priority.

Of the options i looked at this extension seemed to have the most active dev, and has the bonus feature of the diff tool

Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue] Fix #11068 touch #13386 
[Optional Deprecated:]

### refs

- https://github.com/wpilibsuite/sphinxext-rediraffe
- https://pypi.org/project/sphinxext-rediraffe/

### CODE REVIEW

**Comment by phlax** [comment]
trying to figure out how the  actual docs site is built - sphinx is building in `html` mode, but afaict jekyll or somesuch is redirecting from `some/page` to `some/page.html`

given above im reasonably confident that this should fix the redirect problems in #11068 but i cant test 100% until we see it published

**Comment by TheTripleV** [comment]
To run rediraffecheckdiff on PR's, you just need to have both the pr and master available.
This is used in https://github.com/wpilibsuite/frc-docs.

From their ci:
```yaml
  check-redirects:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Checkout master
      run: |
        git fetch origin master --depth=1
    - uses: actions/setup-python@v2
      with:
        python-version: 3.7
    - name: Install Dependencies
      run: |
        pip install -r source/requirements.txt
    - name: Check redirects
      run: |
        make rediraffecheckdiff
```

**Comment by phlax** [comment]
thanks @TheTripleV thats really helpful

**Review by mattklein123 — COMMENTED** [review]
Very cool. Is this something you want to merge right now? Or do you want to wire up more automation?

/wait-any

**Comment by phlax** [comment]
i think if we can land it now, it means we can start moving things without breaking anything.

once we have it landed, i can look at hooking up CI

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
